lib/checkstl.cpp

```diff
@@ -1869,3 +1869,249 @@ void CheckStl::readingEmptyStlContainerError(const Token *tok, const ValueFlow::
 
     reportError(errorPath, value ? (value->errorSeverity() ? Severity::error : Severity::warning) : Severity::style, "reademptycontainer", "$symbol:" + varname +"\n" + errmsg, CWE398, !value);
 }
+
+void CheckStl::useStlAlgorithmError(const Token* tok, const std::string &algoName)
+{
+    reportError(tok, Severity::style, "useStlAlgorithm",
+                "Consider using " + algoName + " algorithm instead of a raw loop.", CWE398, false);
+}
+
+static const Token* singleAssignInScope(const Token* start, unsigned int varid, bool& input)
+{
+    if(start->str() != "{")
+        return nullptr;
+    const Token * endToken = start->link();
+    if(!Token::Match(start->next(), "%var% %assign%"))
+        return nullptr;
+    const Token * endStatement = Token::findsimplematch(start->next(), ";");
+    if(!Token::simpleMatch(endStatement, "; }"))
```

Comment:
with this code you don't seem to handle inner scopes properly:

    {
        if (xy) { dostuff(); }
        /* .. N statements .. */
    }

Reply:
The next condition checks that the brace matches the closing brace, [here](https://github.com/danmar/cppcheck/pull/1352/files/392c7809adde84cba79c8982f55806c1c0fa1c83#diff-b028b3049d71b7931b5ec7906152f7bfR1889):

```cpp
    const Token * endToken = start->link();
    /// ...
    if(!Token::simpleMatch(endStatement, "; }"))
        return nullptr;
    if(endStatement->next() != endToken)
        return nullptr;
```

I am thinking of refactor this as it used often.